error TS2339: Property 'getMetadata' does not exist on type 'typeof Reflect'

Thanks for your reflection lib. I have tried it in my playground project as drop-in replacement for reflect-metadata, but I get the error:

 **error TS2339: Property 'getMetadata' does not exist on type 'typeof Reflect'**

Project: https://github.com/ova2/frontend-tooling-tutorial/tree/master/typescript-playground/dependency-injection-container

**Steps to reproduce:**
1) Clone the project https://github.com/ova2/frontend-tooling-tutorial
2) Go to typescript-playground/dependency-injection-container
2) Remove reflect-metadata from package.json and run "npm install"
3) npm install @abraham/reflection
4)  Replace import 'reflect-metadata' by import '@abraham/reflection' in di-container.ts
5) Run "npm run test" -> error

You will also see the error in your IDE because the proper getMetadata() is not found. Furthermore, I think the return value of this method in your TS definition file should be "any". This is the type in reflect-metadata.

--- a/src/index.test.ts
+++ b/src/index.test.ts
@@ -1,25 +1,31 @@
-import { Reflection as Reflect } from './index';
+import { Reflection } from './index';
 
 test('defineMetadata', () => {
   expect(Reflect.defineMetadata).toBeDefined();
+  expect(Reflect.defineMetadata).toEqual(Reflection.defineMetadata);
 });
 
 test('getMetadata', () => {
   expect(Reflect.getMetadata).toBeDefined();
+  expect(Reflect.getMetadata).toEqual(Reflection.getMetadata);
 });
 
 test('decorate', () => {
   expect(Reflect.decorate).toBeDefined();
+  expect(Reflect.decorate).toEqual(Reflection.decorate);
 });
 
 test('metadata', () => {
   expect(Reflect.metadata).toBeDefined();
+  expect(Reflect.metadata).toEqual(Reflection.metadata);
 });
 
 test('hasOwnMetadata', () => {
   expect(Reflect.hasOwnMetadata).toBeDefined();
+  expect(Reflect.hasOwnMetadata).toEqual(Reflection.hasOwnMetadata);
 });
 
 test('getOwnMetadata', () => {
   expect(Reflect.getOwnMetadata).toBeDefined();
+  expect(Reflect.getOwnMetadata).toEqual(Reflection.getOwnMetadata);
 });

--- a/src/index.ts
+++ b/src/index.ts
@@ -1,12 +1,3 @@
-export const Reflection = Object.assign(Reflect, {
-  decorate,
-  defineMetadata,
-  getMetadata,
-  getOwnMetadata,
-  hasOwnMetadata,
-  metadata,
-});
-
 export type Decorator = ClassDecorator | MemberDecorator;
 export type MemberDecorator = <T>(
   target: Target,
@@ -144,4 +135,27 @@ function createMetadataMap<MetadataValue>(
   const metadataMap = targetMetadata.get(propertyKey) || (new Map<MetadataKey, MetadataValue>());
   targetMetadata.set(propertyKey, metadataMap);
   return metadataMap;
+}
+
+export const Reflection = {
+  decorate,
+  defineMetadata,
+  getMetadata,
+  getOwnMetadata,
+  hasOwnMetadata,
+  metadata,
+};
+
+// TODO: Is this a good approach?
+(window as any).Reflect = Object.assign({}, Reflect, Reflection);
+
+declare global {
+  namespace Reflect {
+    let decorate: typeof Reflection.decorate;
+    let defineMetadata: typeof Reflection.defineMetadata;
+    let getMetadata: typeof Reflection.getMetadata;
+    let getOwnMetadata: typeof Reflection.getOwnMetadata;
+    let hasOwnMetadata: typeof Reflection.hasOwnMetadata;
+    let metadata: typeof Reflection.metadata;
+  }
 }

--- a/src/index.ts
+++ b/src/index.ts
@@ -47,12 +47,12 @@ export function metadata<MetadataValue>(metadataKey: MetadataKey, metadataValue:
   };
 }
 
-export function getMetadata(metadataKey: MetadataKey, target: Target, propertyKey?: PropertyKey) {
-  return ordinaryGetMetadata(metadataKey, target, propertyKey);
+export function getMetadata<MetadataValue>(metadataKey: MetadataKey, target: Target, propertyKey?: PropertyKey) {
+  return ordinaryGetMetadata<MetadataValue>(metadataKey, target, propertyKey);
 }
 
-export function getOwnMetadata(metadataKey: MetadataKey, target: Target, propertyKey?: PropertyKey) {
-  return ordinaryGetOwnMetadata(metadataKey, target, propertyKey);
+export function getOwnMetadata<MetadataValue>(metadataKey: MetadataKey, target: Target, propertyKey?: PropertyKey) {
+  return ordinaryGetOwnMetadata<MetadataValue>(metadataKey, target, propertyKey);
 }
 
 export function hasOwnMetadata(metadataKey: MetadataKey, target: Target, propertyKey?: PropertyKey): boolean {
@@ -95,13 +95,13 @@ function ordinaryDefineOwnMetadata<MetadataValue>(
     .set(metadataKey, metadataValue);
 }
 
-function ordinaryGetMetadata(
+function ordinaryGetMetadata<MetadataValue>(
   metadataKey: MetadataKey,
   target: Target,
   propertyKey?: PropertyKey,
-): Function | undefined {
-  return !!ordinaryGetOwnMetadata(metadataKey, target, propertyKey)
-    ? ordinaryGetOwnMetadata(metadataKey, target, propertyKey)
+): MetadataValue | undefined {
+  return !!ordinaryGetOwnMetadata<MetadataValue>(metadataKey, target, propertyKey)
+    ? ordinaryGetOwnMetadata<MetadataValue>(metadataKey, target, propertyKey)
     : Object.getPrototypeOf(target)
     ? ordinaryGetMetadata(metadataKey, Object.getPrototypeOf(target), propertyKey)
     : undefined;

--- a/tslint.json
+++ b/tslint.json
@@ -15,7 +15,8 @@
       ["String", "Avoid using the `String` type. Did you mean `string`?"],
       ["Symbol", "Avoid using the `Symbol` type. Did you mean `symbol`?"]
     ],
-    "no-shadowed-variable": false
+    "no-shadowed-variable": false,
+    "no-namespace": false
   },
   "rulesDirectory": []
 }

